### PR TITLE
Add timing metrics for case seach

### PR DESCRIPTION
**Pull request by esoergel**
## Product Description
Admin-only view to see case search performance breakdown.  Works best with the `LongCaseSearchRequest` log in sentry

image

## Technical Summary
This initializes a `TimingContext` object when executing a case search request and keeps track of various sections of the process.  Turns out at least in some instances it's not just the ES query that's slow, but also building the query (and executing secondary queries), wrapping cases, serializing cases, fetching an app (seems like that shouldn't be happening here), and enriching the data with related cases.  I want to see what this looks like on production, since the circumstances there are pretty different than staging.

The timing code is not necessary in normal case search queries, only when initiating through this admin view, but I don't know of a better way to do this, and I can't imagine this having an appreciable performance impact.  It does add a bit of boilerplate to the code, which I don't like, and I'd support removing this when we are no longer as focused on case search performance.

### Safety story
The parts of this that involve production code is relatively small, and that's the critical part.  The bulk of the changeset is admin-only, and I'm not concerned about errors there.  This has been on staging for about a week and I've tested it there without issue.

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/case_search/utils.py [hunk 1]
```diff
@@ -328,11 +373,13 @@ def get_and_tag_related_cases(helper, app_id, case_types, cases,
...
-    for case in results:
-        _tag_is_related_case(case)
+    with helper.timing_context('_tag_is_related_case'):
+        for case in results:
+            _tag_is_related_case(case)
     return results
 
 
+@time_function()
 def get_related_cases_result(helper, app, case_types, source_cases, include_all_related_cases):
```
Comment (Jtang-1) [inline corehq/apps/case_search/utils.py:383]: This could either result in `_get_all_related_cases` OR `_get_search_detail_path_defined_cases` and `_get_child_cases_referenced_in_app`. Would it also be helpful to get the label the timing context for those parts?

**Review by Jtang-1 — APPROVED** [review]
I focused more on the non-admin page related changes. But this is super cool! Just one question

**Comment by snopoke** [comment]
I haven't looked at the code but I definitely support improving our visibility into core platform components like case search. I can imagine this being generally useful in future, even if we aren't specifically focusing on case search performance, though it would be more useful if app builders could somehow get access to this info (like having a debug mode in web apps or something) - but that's getting a bit off topic.

One thing I notice from the screenshot is that there's no indication of how many times a function was called. For queries that use subcase / ancestor functions it would be useful to have that split out of get some insight into it.

I also wonder if being able to see the actual ES queries that were executed would ever be useful.

Just some thought....

File: corehq/apps/case_search/static/case_search/js/profile_case_search.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    var caseSearchModel = function () {
+        var self = {};
+        self.appId = ko.observable();
+        self.requestDict = ko.observable();
+        self.results = ko.observable();
+        self.browserTime = ko.observable();
+        self.searchButtonIcon = ko.observable("fa fa-search");
+        self.exampleRequestDict = JSON.stringify({
+            "request_dict": {
+                "_xpath_query": ["name: 'Ethan'", "subcase-exists(parent, name='Matilda')"],
```
Comment (snopoke) [inline corehq/apps/case_search/static/case_search/js/profile_case_search.js:23]: ❤️

File: corehq/apps/case_search/templates/case_search/case_search.html [hunk 3]
```diff
@@ -87,26 +87,18 @@ <h3>{% trans "Search Parameters" %}</h3>
...
+                  <span class='btn btn-danger' data-bind="click: $parent.removeXPath">
+                    <i class="fa-regular fa-trash-can"></i> {% trans "Delete" %}
+                  </span>
                 </div>
+                <button class="btn btn-default" id="add-xpath" data-bind="click: addXPath">
+                  <i class="fa fa-plus"></i> {% trans "Add XPath" %}
+                </button>
               </div>
             </div>
-            {% if query_additions %}
```
Comment (snopoke) [inline corehq/apps/case_search/templates/case_search/case_search.html:95]: Just noting that references to this were removed here: https://github.com/dimagi/commcare-hq/pull/28278

File: corehq/apps/case_search/templates/case_search/case_search.html [hunk 4]
```diff
@@ -87,26 +87,18 @@ <h3>{% trans "Search Parameters" %}</h3>
...
+                <div class="form-group" style="width:100%" data-bind="foreach: xpath_expressions" >
+                  <textarea style="min-width: 50%" spellcheck="false" class="form-control vertical-resize"
+                    name="xpath" type="text" placeholder="e.g: name = 'tyrion' or nickname='the imp'" data-bind="value: xpath"> </textarea>
+                  <span class='btn btn-danger' data-bind="click: $parent.removeXPath">
+                    <i class="fa-regular fa-trash-can"></i> {% trans "Delete" %}
+                  </span>
                 </div>
+                <button class="btn btn-default" id="add-xpath" data-bind="click: addXPath">
+                  <i class="fa fa-plus"></i> {% trans "Add XPath" %}
+                </button>
```
Comment (snopoke) [inline corehq/apps/case_search/templates/case_search/case_search.html:99]: These changes seem unrelated to the profiling, are they just a bonus?

File: corehq/apps/case_search/tests/test_models.py [hunk 5]
```diff
@@ -85,15 +84,10 @@ def test_extract_criteria_config(self, case_type, data_registry, custom_related_
...
-            continue
-        if isinstance(value, list):
-            request_dict.setlist(key, value)
-        else:
-            request_dict[key] = value
-    return request_dict
+    return {
+        key: (value if isinstance(value, list) else [value])
+        for key, value in params.items() if value is not None
+    }
```
Comment (snopoke) [inline corehq/apps/case_search/tests/test_models.py:90]: do none of the uses rely on any of the methods of `MultiValueDict` e.g. `getlist`?

**Review by snopoke — APPROVED** [review]

File: corehq/apps/case_search/tests/test_models.py (hunk 5 above)
Comment (esoergel) [reply corehq/apps/case_search/tests/test_models.py:90]: Nope, it's only used in one place, which converts it to a normal dict.  I moved that coercion up the stack to make it easier to initialize in other contexts.
https://github.com/dimagi/commcare-hq/blob/ef722b033b9dfe1dc15b4129915806a7152ea183/corehq/apps/case_search/models.py#L199-L200

**Comment by esoergel** [comment]
> it would be more useful if app builders could somehow get access to this info (like having a debug mode in web apps or something)

That would be really cool.  We could also store the timing information of real requests over some threshold and view it like a report later - that'd be super helpful.

> One thing I notice from the screenshot is that there's no indication of how many times a function was called. For queries that use subcase / ancestor functions it would be useful to have that split out of get some insight into it.
> I also wonder if being able to see the actual ES queries that were executed would ever be useful.

I agree with both of these - I'd like to explore that next, though I think they'll be a little harder, as you need access to the `timing_context` deeper down the stack.  I was thinking about maybe making some `profiler` object to hold the timing context, the total counts, and things like actual JSON of the queries run, but I'll have to figure out a way to make it accessible that's not too crufty.

**Review by Jtang-1 — APPROVED** [review]
